win32/src/gballoc_hl_metrics.c

```diff
@@ -0,0 +1,435 @@
+    { "Bucket [268435456-536870911]", 268435456, 536870911 },
+    { "Bucket [536870912-1073741823]", 536870912, 1073741823 },
+    { "Bucket [1073741824-2147483647]", 1073741824, 2147483647 },
+    { "Bucket [2147483648-4294967295]", 2147483648, 4294967295 }
+};
+
+typedef struct LATENCY_BUCKET_TAG
+{
+    volatile LONG64 latency_sum;
+    volatile LONG latency_min;
+    volatile LONG latency_max;
+    volatile LONG count;
+} LATENCY_BUCKET;
+
+static LATENCY_BUCKET malloc_latency_buckets[GBALLOC_LATENCY_BUCKET_COUNT];
+static LATENCY_BUCKET calloc_latency_buckets[GBALLOC_LATENCY_BUCKET_COUNT];
+static LATENCY_BUCKET realloc_latency_buckets[GBALLOC_LATENCY_BUCKET_COUNT];
+static LATENCY_BUCKET free_latency_buckets[GBALLOC_LATENCY_BUCKET_COUNT];
+
+static size_t determine_latency_bucket_for_size(size_t size)
+{
+    size_t bucket = 0;
+
+#if SIZE_MAX != UINT32_MAX
+    if (size > UINT32_MAX)
+    {
+        bucket = GBALLOC_LATENCY_BUCKET_COUNT - 1;
+    }
+    else
+#endif
+    {
+        // start at 512
+        size >>= 9;
+        while (size != 0)
+        {
+            bucket++;
+            size >>= 1;
+        }
+    }
+
+    return bucket;
+}
+
+static void init_latency_bucket(LATENCY_BUCKET* latency_bucket)
+{
+    (void)InterlockedExchange(&latency_bucket->count, 0);
+    (void)InterlockedExchange64(&latency_bucket->latency_sum, 0);
+    (void)InterlockedExchange(&latency_bucket->latency_min, LONG_MAX);
+    (void)InterlockedExchange(&latency_bucket->latency_max, 0);
+}
+
+static void internal_init_latency_counters(void)
+{
+    size_t i;
+
+    for (i = 0; i < GBALLOC_LATENCY_BUCKET_COUNT; i++)
+    {
+        init_latency_bucket(&malloc_latency_buckets[i]);
+        init_latency_bucket(&calloc_latency_buckets[i]);
+        init_latency_bucket(&realloc_latency_buckets[i]);
+        init_latency_bucket(&free_latency_buckets[i]);
+    }
+}
+
+static void internal_copy_latency_data(GBALLOC_LATENCY_BUCKETS* latency_buckets_out, LATENCY_BUCKET* source_latency_buckets)
+{
+    size_t i;
+
+    for (i = 0; i < GBALLOC_LATENCY_BUCKET_COUNT; i++)
+    {
+        latency_buckets_out->buckets[i].count = InterlockedAdd(&source_latency_buckets[i].count, 0);
+        latency_buckets_out->buckets[i].latency_avg = (latency_buckets_out->buckets[i].count == 0) ? 0 : (double)InterlockedAdd64(&source_latency_buckets[i].latency_sum, 0) / latency_buckets_out->buckets[i].count;
+        latency_buckets_out->buckets[i].latency_min = InterlockedAdd(&source_latency_buckets[i].latency_min, 0);
+        latency_buckets_out->buckets[i].latency_max = InterlockedAdd(&source_latency_buckets[i].latency_max, 0);
+    }
+}
+
+static int wasInitialized = 0;
+
+int gballoc_hl_init(void* hl_params, void* ll_params)
+{
+    (void)hl_params;
+    int result;
+
+    /* Codes_SRS_GBALLOC_HL_METRICS_01_001: [ If the module is already initialized, gballoc_hl_init shall fail and return a non-zero value. ]*/
+    if (wasInitialized != 0)

```

Comment:
>wasInitialized [](start = 8, length = 14)

I think the spec should mention that init/deinit are not thread-safe

Reply:
None of the init/deinit are. Added here and everywhere that the function is not thread-safe.

---
In reply to: [459659999](https://github.com/Azure/azure-c-pal/pull/14#discussion_r459659999) [](ancestors = 459659999)